src/arena.py

```diff
@@ -1 +1,156 @@
-WORKS = True
+import json
+import re
+import random
+
+import falcon
+from wsgiref import simple_server
+from typing import List
+from json import JSONEncoder
+
+from dataclasses import dataclass
+
+
+class Combatant:
+
+    def __init__(self, name: str, hp: int, damage: str):
+        self.name = name
+        self.health = hp
+        self.__damage = self.Damage(damage)
+        self.last_damage = 0
+        self.last_health = hp
+
+    def attack(self, other: 'Combatant'):
+        self.last_damage = self.__damage.draw()
+        other.__get_attacked(self.last_damage)
+
+    def __get_attacked(self, damage: int):
+        self.last_health = self.health
+        self.health -= damage
+        if self.health < 0:
+            self.health = 0
+
+    def is_dead(self) -> bool:
+        return self.health == 0
+
+    class Damage:
+        __dice_roll_number: int
+        __dice_sides_number: int
+        __attack_bonus: int
+
+        def __init__(self, damage: str):
+            # XdY[ + Z]
+            m = re.search(r"^(\d+)d(\d+)( \+ (\d+))?$", damage)
+            if m:
+                self.__dice_roll_number = int(m.group(1))
+                self.__dice_sides_number = int(m.group(2))
+                z = m.group(4)
+                if z:
+                    self.__attack_bonus = int(z)
+                else:
+                    self.__attack_bonus = 0
+
+        def draw(self):
+            result = 0
+            for i in range(self.__dice_roll_number):
+                result += random.randint(1, self.__dice_sides_number)
+            result += self.__attack_bonus
+            return result
+
+
+Combatants = List[Combatant]
+
+
+class Arena:
+    c1: Combatant
+    c2: Combatant
+    max_rounds: int
+
+    def __init__(self, c1: Combatant, c2: Combatant, max_rounds: int):
+        self.c1 = c1
+        self.c2 = c2
+        self.max_rounds = max_rounds
+
+    def fight(self) -> 'BattleResult':
+        rounds: 'BattleResult.Rounds' = []
+        victory: 'BattleResult.Victory'
+        round_number: int = 1
+        while not self.c1.is_dead() and not self.c1.is_dead() and round_number <= self.max_rounds:
+            rounds.append(self.next_round(round_number, self.c1, self.c2))
+            self.c1, self.c2 = self.c2, self.c1
+            round_number += 1
+        else:
```

Comment:
`while-else` makes sense only if you use `break` inside `while`.
In this case you can just skip oit and put the code next to while.
On the other hand - this would make sense:
```python
for round in range(self.max_rounds):
     ...
     if self.c1.is_dead of self.c2.is_dead:
         break
else:
    # No ko
```

You can read more about it here: https://realpython.com/python-while-loop/#the-else-clause

Reply:
ok